ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -937,6 +890,323 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+         --   then there would be no intersection anymore).
+         -- + This means the new anchor point must be between the old anchor
+         --   point and the new intersection point.
+         -- + Since we know both the old anchor point and the new
+         --   intersection point exist on their fragment, the new anchor
+         --   point must also.
+         Nothing -> error
+             "anchor point must be on candidate fragment if they intersect"
+         Just (_, trimmedCandidateFrag) -> return $ NewIntersection $
+             assertKnownIntersectionInvariants (configConsensus cfg) $
+               KnownIntersectionState {
+                   ourFrag                 = ourFrag'
+                 , theirFrag               = trimmedCandidateFrag
+                 , theirHeaderStateHistory = trimmedHeaderStateHistory'
+                 , mostRecentIntersection  = castPoint intersection
+                 }
+           where
+             -- We trim the 'HeaderStateHistory' to the same size as our
+             -- fragment so they keep in sync.
+             trimmedHeaderStateHistory' =
+               HeaderStateHistory.trim
+                 (AF.length trimmedCandidateFrag)
+                 theirHeaderStateHistory
+
+      | otherwise ->
+        -- No more intersection with the current chain
+        return LostIntersection
+  where
+    ChainDbView {
+        getCurrentChain
+      } = cdbView
+    KnownIntersectionState {
+        theirFrag
+      , theirHeaderStateHistory
+      , ourFrag
+      } = kis
+
+{-------------------------------------------------------------------------------
+  Acquiring a ledger view for RollForward validation
+-------------------------------------------------------------------------------}
+
+data LedgerViewCheck blk =
+    -- | The upstream chain no longer intersects with our current chain because
+    -- our current chain changed in the background.
+    NoLongerIntersects
+    -- | The upstream chain still intersects with our chain, return the latest
+    -- 'KnownIntersectionState' and the 'LedgerView' necessary to validate the
+    -- RollForward header.
+  | LedgerViewAcquired
+      (KnownIntersectionState blk)
+      (Ticked (LedgerView (BlockProtocol blk)))
+
+-- | The peer has sent a RollForward header but their chain is not dense enough
+-- for us to validate the header in the usual way (via forecasting). We'll have
+-- to pre-emptively fetch some of their blocks in order to do so.
+--
+-- This is the state type for the logic requests fetchs and processes the
+-- fetched blocks.
+data LowDensity blk
+  = LowDensity
+      !(KnownIntersectionState blk)
+      -- ^ The 'mostRecentIntersection' via which we obtain the
+      -- initially-unadvanced 'LedgerState' (whose 'LedgerView' forecast did
+      -- not reach the slot of the RollForward header).
+      !(LedgerState blk)
+      -- ^ The advanced ledger state so far.
+      ![RealPoint blk]
```

Comment:
We don't like to put a bang on a list, as it is misleading, it just forces it to `_ : _` or `[]`. We likely have a few of such bangs somewhere, but IIRC, we decided to leave them out in the future. (It feels inconsistent to leave it out :slightly_smiling_face:).

Reply:
Remove the bang